fastlane/Fastfile

```diff
@@ -424,16 +355,125 @@ platform :android do
     version = release_version_current
 
     # Wrap up
-    remove_branch_protection(repository: GH_REPOSITORY, branch: "release/#{version}")
-    set_milestone_frozen_marker(repository: GH_REPOSITORY, milestone: version, freeze: false)
-    create_new_milestone(repository: GH_REPOSITORY, need_appstore_submission: true)
-    close_milestone(repository: GH_REPOSITORY, milestone: version)
+    remove_branch_protection(repository: GITHUB_REPO, branch: "release/#{version}")
+    set_milestone_frozen_marker(repository: GITHUB_REPO, milestone: version, freeze: false)
+    create_new_milestone(repository: GITHUB_REPO, need_appstore_submission: true)
+    close_milestone(repository: GITHUB_REPO, milestone: version)
 
     push_to_git_remote(tags: false)
     trigger_release_build(branch_to_build: "release/#{version}")
     create_backmerge_pr
   end
 
+  # This lane publishes a release on GitHub and creates a PR to backmerge the current release branch into the next release/ branch
+  #
+  # @param [Boolean] skip_confirm (default: false) If set, will skip the confirmation prompt before running the rest of the lane
+  #
+  lane :publish_release do |skip_confirm: false|
+    ensure_git_status_clean
+    ensure_git_branch(branch: '^release/')
+
+    version_number = release_version_current
+
+    current_branch = "release/#{version_number}"
+    next_release_branch = "release/#{release_version_next}"
+
+    UI.important <<~PROMPT
+      Publish the #{version_number} release. This will:
+      - Publish the existing draft `#{version_number}` release on GitHub
+      - Which will also have GitHub create the associated git tag, pointing to the tip of the branch
+      - If the release branch for the next version `#{next_release_branch}` already exists, backmerge `#{current_branch}` into it
+      - If needed, backmerge `#{current_branch}` back into `#{DEFAULT_BRANCH}`
+      - Delete the `#{current_branch}` branch
+    PROMPT
+    UI.user_error!("Terminating as requested. Don't forget to run the remainder of this automation manually.") unless skip_confirm || UI.confirm('Do you want to continue?')
+
+    UI.important "Publishing release #{version_number} on GitHub"
+
+    publish_github_release(
+      repository: GITHUB_REPO,
+      name: version_number
+    )
+
+    create_backmerge_pr
```

Comment:
See @iangmaia 's [comment on the original PR](https://github.com/Automattic/pocket-casts-android/pull/3371#discussion_r1888281340) about this.

 - In most cases there won't be anything to merge from `release/N->main` at that stage so creation of such a PR would be expected to be skipped by the `create_release_backmerge_pull_request` action… but given PCAndroid uses "Squash Merge", the action will detect that some commits from `release/N` are not present in `main` and thus still create such a `release/N->main` backmerge PR anyway.
 - Given that we don't expect any last-minute commits having landed in `release/N` branch _**after**_ the `finalize_release` happened earlier in the scenario, in practice it should be OK 99% of the time to skip the creation of this `release/N->main` backmerge PR. There's still a risk that a late commit that didn't warrant a resubmission (e.g. last-minute tooling fix landed after `finalize_release` but not requiring a new run of `finalize_release` after it landed because it only changed tooling, not the binary being submitted) would be lost though, but that might be an acceptable balance too?
 - It's important to note though that, regardless of all the above, we still want to ensure that there's a `release/N->release/N+1` backmerge PR being created to make sure we include all the latest fixes of version N in version N+1 for those late fixes that might have landed after the code freeze of version `N+1` (and the creation of `release/N+1` branch) already happened.


As a result, we might want to adjust this call to only backmerge to the next `release/N+1` branch (if it already exists at that time, which should be the case given the scenario timing)… but not to `main`?

---

Ideally we should probably find a way to improve our `create_release_backmerge_pull_request` action in `release-toolkit` so it can support the case of repos that are using squash-merges, so that its implementation logic to detect if such a PR is necessary would not be based on "if there are commits in the `release/*` branch not present in the merge-base of `release/*` and `main`", but instead based on the diff being empty or not (worth pondering if we should then do such a comparison on the two-dots vs three-dots diff and consider any potential edge cases though)

In the meantime, since the current Release Scenario never mentioned creating a backmerge PR amongst the tasks of the "Release" milestone, maybe it'd be just as simple to remove that call from `publish_release`, to keep the current status quo? 🤔

Reply:
> Ideally we should probably find a way to improve our create_release_backmerge_pull_request action in release-toolkit so it can support the case of repos that are using squash-merges

I agree. I vaguely remember one of my first implementations was using the actual diff, but iinm using _only_ the diff had its own problems. Perhaps we could combine the current implementation with that approach as well 🤔 

Also to ponder, GitHub itself doesn't seem to use such approach and just creates the PR 🤷 

> In the meantime, since the current Release Scenario never mentioned creating a backmerge PR amongst the tasks of the "Release" milestone, maybe it'd be just as simple to remove that call from `publish_release`, to keep the current status quo? 🤔

Yep, I've decided together with the team to follow this approach for WPAndroid / WPiOS for now. Additionally, they also argued that creating a PR from this step can be somewhat unexpected for the release manager, but I guess that could be solved with either a separate step or better communication.

Reply:
> Additionally, they also argued that creating a PR from this step can be somewhat unexpected for the release manager, but I guess that could be solved with either a separate step or better communication.

I can see how it can be confuising to have a backmerge PR `release/N->trunk` be created at that stage if no new commit has landed in `release/N` since the submission, but yet the action creates a PR nonetheless because of the repo using squash-merge and the action thus detecting commits from `release/N` missing from `trunk`.

But if we solve this to support the use case of repos using squash-merge (maybe just checking for the diff being non-empty in those cases indeed?), then at that point I think we should still keep the `publish_release` lane being the one doing the backmerge PRs, including because it's also the lane that ensures that the `release/N` branch is deleted once the tag has been created by the GitHub Release.

We can't delete the `release/*` branch until we've made sure we created a git tag (by way of publishing the GitHub release) first, and we can't delete the `release/*` branch until we ensured that everything that was in that `release/*` branch has been propagated to other branches (`trunk` and `release/N+1`) first (either by a merge-commit or by a squash-merge), otherwise for the very rare cases where we might have added a commit post-`finalize_release` (e.g. tooling fix or similar) we risk losing that fix forever instead of it being propagated to next versions. So making this backmerge PR be done in a separate `Task` in the release scenario might not be practically possible as the `release/*` branch would have already been deleted by the previous task calling `publish_release` at that point.

But as long as we manage to make the `create_release_backmerge_pull_request` action properly detect when a PR is not necessary on repos using squash-merge like it does for repos using merge-commits, in 95% of cases there won't be a PR created at `publish_release` stage anymore so that won't create some unexpected surprise for the release manager, all while ensuring that in the rare cases where it would be necessary to create the PR to ensure not to lose post-submission (tooling?) fixes, it is still done.

Reply:
> as long as we manage to make the `create_release_backmerge_pull_request` action properly detect when a PR is not necessary on repos using squash-merge like it does for repos using merge-commits

Yeah, I think that's the main point: making the action work properly for repos with a squash merge policy.

The PR(s) creation during release publishing (or eventual breakdown into different Rv2 tasks, or a better task description to the release manager, etc) can be a separate discussion.